client/components/swipeable/index.js

```diff
@@ -0,0 +1,296 @@
+	// save callback to a ref so that it doesn't need to be a dependency of other hooks
+	const savedUpdateLayout = useRef();
+	useEffect( () => {
+		savedUpdateLayout.current = updateLayout;
+	}, [ updateLayout ] );
+
+	// fire when the `currentPageIndex` parameter changes
+	useEffect( () => {
+		if ( ! enabled ) {
+			return;
+		}
+
+		savedUpdateLayout.current();
+	}, [ enabled, currentPageIndex ] );
+
+	// fire when the window resizes
+	useEffect( () => {
+		if ( ! enabled ) {
+			return;
+		}
+
+		const onResize = () => savedUpdateLayout.current( false );
+		window.addEventListener( 'resize', onResize );
+		window.addEventListener( 'orientationchange', onResize );
+		return () => {
+			window.removeEventListener( 'resize', onResize );
+			window.removeEventListener( 'orientationchange', onResize );
+		};
+	}, [ enabled ] );
+}
+
+function getDragPositionAndTime( event ) {
+	const { timeStamp } = event;
+	if ( event.hasOwnProperty( 'clientX' ) ) {
+		return { x: event.clientX, y: event.clientY, timeStamp };
+	}
+
+	if ( event.targetTouches[ 0 ] ) {
+		return {
+			x: event.targetTouches[ 0 ].clientX,
+			y: event.targetTouches[ 0 ].clientY,
+			timeStamp,
+		};
+	}
+
+	const touch = event.changedTouches[ 0 ];
+	return { x: touch.clientX, y: touch.clientY, timeStamp };
+}
+
+export const Swipeable = ( {
+	hasDynamicHeight = false,
+	children,
+	currentPage = 0,
+	onPageSelect,
+	pageClassName,
+	containerClassName,
+	...otherProps
+} ) => {
+	const [ pageWidth, setPageWidth ] = useState( null );
+	const [ swipeableArea, setSwipeableArea ] = useState();
+	const isRtl = useRtl();
+
+	const [ pagesStyle, setPagesStyle ] = useState( {
+		transform: `translate3d(0px, 0px, 0px)`,
+		transitionDuration: `300ms`,
+	} );
+	const [ dragStartData, setDragStartData ] = useState( null );
+
+	const pagesRef = useRef();
+	const resizeObserverRef = useRef();
+	const numPages = Children.count( children );
+
+	const getWidth = useCallback( () => {
+		return resizeObserverRef.current?.getBoundingClientRect().width;
+	}, [ resizeObserverRef ] );
+
+	const getPagesWidth = useCallback( () => {
+		if ( ! pageWidth ) {
+			return null;
+		}
+		return pageWidth * numPages;
+	}, [ pageWidth, numPages ] );
+
+	const getOffset = useCallback(
+		( index ) => {
+			const offset = pageWidth * index;
+			return isRtl ? offset : -offset;
+		},
+		[ isRtl, pageWidth ]
+	);
+
+	const updateEnabled = hasDynamicHeight && numPages > 1;
+
+	useUpdateLayout( updateEnabled, currentPage, ( hasTransitionDuration = true ) => {
+		const offset = getOffset( currentPage );
+		const transform = {
+			transform: `translate3d(${ offset }px, 0px, 0px)`,
+			transitionDuration: hasTransitionDuration ? `300ms` : `0ms`,
+		};
+
+		let height = { height: null };
+		const targetHeight = pagesRef.current?.querySelector( '.is-current' )?.offsetHeight;
+
+		if ( targetHeight && pagesStyle?.height !== targetHeight ) {
+			height = { height: null };
+		}
+		setPageWidth( getWidth() );
```

Comment:
Since the `getWidth()` function is only invoked here, we could also remove that function and replace this line with:

```suggestion
		setPageWidth( resizeObserverRef.current?.getBoundingClientRect().width );
```